omegaconf/listconfig.py

```diff
@@ -102,17 +103,32 @@ def _validate_set(self, key: Any, value: Any) -> None:
 
         target_type = self._metadata.element_type
         value_type = OmegaConf.get_type(value)
-        if is_structured_config(target_type):
+
+        def raise_validation_error(
+            value_type: Any, target_type: Any, value: Any
+        ) -> None:
+            msg = (
+                f"Invalid type assigned : {type_str(value_type)} is not a "
+                f"subclass of {type_str(target_type)}. value: {value}"
+            )
+
+            raise ValidationError(msg)
+
+        if is_structured_config(target_type) or (
+            target_type is not Any and isinstance(value, ValueNode)
+        ):
             if (
                 target_type is not None
                 and value_type is not None
                 and not issubclass(value_type, target_type)
             ):
-                msg = (
-                    f"Invalid type assigned : {type_str(value_type)} is not a "
-                    f"subclass of {type_str(target_type)}. value: {value}"
-                )
-                raise ValidationError(msg)
+                raise_validation_error(value_type, target_type, value)
+        if (
+            is_structured_config(value_type)
+            and not is_structured_config(target_type)
+            and target_type not in (Any, None)
+        ):
+            raise_validation_error(value_type, target_type, value)
```

Comment:
this block is confused. can you reorganized it to be clearer? (adding a comment about which failure case each condition is detecting will also help).